### PR TITLE
Auto-generate options in README from field attributes

### PR DESCRIPTION
**Pull request by squiddy**
This implements a proc macro to derive metadata from the structs that define all available options. I'm opening this pull request primarly to get feedback on whether this is a sensible approach.

`src/settings/options.rs` shows how the attributes are used to add additional information to options.
`ruff_dev/src/generate_options.rs` shows how that is used to autogenerate the options in the readme.

I'm using macros here mostly to learn about them (which is why the macro implementation is pretty rough), but also because I think they fit well here. And they might help us structuring all the lints better, similar to clippy does it?

- [x] macro: cleanup code
- [x] macro: proper error handling
- [x] doc: complete root options
- [x] doc: complete flake8_annotations
- [x] doc: complete flake8_bugbear
- [x] doc: complete flake8_quotes
- [x] doc: complete flake8_tidy_imports
- [x] doc: complete eisort
- [x] doc: complete mccabe
- [x] doc: complete pep8_naming
- [x] doc: complete pyupgrade

### CODE REVIEW

**Comment by charliermarsh** [comment]
This is really interesting! Learning a lot just from reading it (I'm not a macro expert by any means).

**Comment by charliermarsh** [comment]
I think this makes a lot of sense. I have some small nits on the code but I'd def encourage you to follow this through to its conclusion (e.g., we need to handle all the various `Options` structs, support multiple sections, and so on), and I'm happy to merge it as soon as it generates the existing docs.

**Comment by squiddy** [comment]
Thanks for the feedback.

The macro code itself is still a mess, but I started attributing more options. Groups are also supported now.

**Comment by charliermarsh** [comment]
If you want to focus on macro clean-up, I can commit to translating all of the remaining options to use that syntax once the PR is good to go.

**Comment by charliermarsh** [comment]
(I don't mind the grunt work.)

**Comment by squiddy** [comment]
Oh yeah, that would be great. Might also be a good opportunity to spot things that I could still improve.

I think I made some good steps with the macro cleanup already (learning on the go), will push that later.

**Comment by charliermarsh** [comment]
Will finish this tonight. I'd like to tweak the ordering of the options. It'd be nice, too, if `per-file-ignores` could appear at the top, despite being at the bottom of the struct for dumb TOML reasons.

File: README.md [hunk 1]
```diff
@@ -1283,6 +1283,26 @@ Summary
...
+
+**Default value**: `[]`
+
+**Type**: `Vec<char>`
+
+**Example usage**:
+
+```toml
+[tool.ruff]
+# Allow minus-sign (U+2212), greek-small-letter-rho (U+03C1), and greek-small-letter-alpha (U+03B1),
```
Comment (eddyg) [inline README.md:1299]: this should say: 

> ... and the asterisk-operator (U+2217)

(assuming you want to use `"*"` and `"∗"` in the example) 😉

Comment (charliermarsh) [reply README.md]: Thanks!

**Pull request merged**